src/unittests/simple_commander.py

```diff
@@ -0,0 +1,94 @@
+import unittest
+import math
+import time
+
+from src.simple_commander.controllers.main import GameController
+from src.simple_commander.controllers.main import Bullet
+from src.simple_commander.controllers.main import Hero
+from src.simple_commander.controllers.main import Invader
+from src.simple_commander.controllers.main import DEFAULT_SPEED
+
+
+class MainControllerTestCase(unittest.TestCase):
+    def setUp(self):
+        self.game_object = GameController(100, 100, 2)
+        self.max_height = self.game_object.game_field['height']
+        self.max_width = self.game_object.game_field['width']
+
+    def tearDown(self):
+        del self
+
+    def test_game_field_width(self):
+        self.assertEqual(self.game_object.game_field['width'], 100, 'incorrect game field width')
+
+    def test_game_field_height(self):
+        self.assertEqual(self.game_object.game_field['height'], 100, 'incorrect game field height')
+
+    def test_game_field_units_length(self):
+        self.assertEqual(len(self.game_object.units), 3, 'incorrect length of units')
+
+    def test_bullet_create(self):
+        self.game_object.fire(self.game_object.units[0])
+        self.assertEqual(len(self.game_object.units), 4, 'Bullet was not created')
+
+    def test_hero_start_speed(self):
+        self.assertEqual(self.game_object.units[0].speed, 0, 'Incorrect initial speed')
+        self.assertEqual(self.game_object.units[1].speed, DEFAULT_SPEED, 'Incorrect initial speed')
+
+    def test_hero_speed_after_change(self):
+        self.game_object.units[0].change_speed(5)
+        self.assertEqual(self.game_object.units[0].speed, 5, 'Incorrect initial speed')
+        self.assertEqual(self.game_object.units[1].speed, DEFAULT_SPEED, 'Incorrect initial speed')
+
+    def test_hero_angle(self):
+        hero = Hero(20, 20, 100)
+        self.assertEqual(hero.angle, 100, 'Incorrect initial angle')
+
+    def test_hero_angle_after_rotate(self):
+        hero = Hero(20, 20, 100)
+        hero.rotate(20)
+        self.assertEqual(hero.angle, 120, 'Incorrect angle after rotate')
+
+    #  Bullet coordinates after initial x = 0, y = 0, angle = 90:
+    #    1. after 0 seconds --- step = 0, x = 0, y = 0
+    #    2. after 1 seconds --- step = 1, x = -2, y = 0
+    #    3. after 2 seconds --- step = 2, x = -4, y = -11
+    #    4. after 3 seconds --- step = 3, x = -7, y = -31
+    #  Coordinate will be the same as in invader!! Both units will be dead !!
+    def test_compute_new_coordinate(self):
+        hero = Hero(0, 0, 90)
+        bullet = Bullet(hero)
+        while bullet.step < 3:
+            bullet.compute_new_coordinate(self.game_object.game_field)
+        self.assertEqual(bullet.x, -7, 'Bullet coordinate x is incorrect!')
+        self.assertEqual(bullet.y, -31, 'Bullet coordinate y is incorrect!')
```

Comment:
You hero is positioned on 90 degrees and it mean that will move by X axis, Y must be stable. Though why do you check Y for changes and even in reverse direction? I my mind the new coordinates must be (5, 0) for stale speed 5 and step duration 1 sec